src/eventhandling/hovering.rs

```diff
@@ -0,0 +1,204 @@
+use amethyst::{
+    core::{
+        nalgebra::{try_convert, Isometry, Isometry3, Translation3},
+        GlobalTransform,
+    },
+    ecs::prelude::{
+        Component, DenseVecStorage, Entities, Entity, Join, Read, ReadStorage, System, World,
+        Write, WriteStorage,
+    },
+    renderer::{Material, TextureHandle},
+    shrev::EventChannel,
+};
+
+use eventhandling::MouseRay;
+use ncollide3d::shape::Shape;
+
+pub struct HoverInteractionSystem;
+
+impl<'a> System<'a> for HoverInteractionSystem {
+    type SystemData = (
+        Entities<'a>,
+        Read<'a, MouseRay>,
+        ReadStorage<'a, GlobalTransform>,
+        WriteStorage<'a, HoverHandlerComponent>,
+        Write<'a, Hovered>,
+        Write<'a, EventChannel<HoverEvent>>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mouse_ray, transforms, mut hover_handlers, mut hovered, mut hover_channel): Self::SystemData,
+    ) {
+        let mut nearest_dist = None;
+        let mut nearest_entity = None;
+        for (entity, hover_handler, transform) in
+            (&*entities, &mut hover_handlers, &transforms).join()
+        {
+            if let Some(collision_distance) = {
+                // the mesh model has its pivot point on the bottom, the collider have their pivots in the middle.
+                // to adjust the position we need to move the collision shapes a bit up
+                let offset: Translation3<f32> = Translation3::new(
+                    0.0,
+                    hover_handler
+                        .bounding_box()
+                        .aabb(&Isometry::identity())
+                        .half_extents()
+                        .y,
+                    0.0,
+                );
+                let mut translation: Isometry3<f32> = try_convert(transform.0).unwrap();
+                translation.append_translation_mut(&offset);
+
+                hover_handler
+                    .bounding_box()
+                    .as_ray_cast()
+                    .unwrap()
+                    .toi_with_ray(&translation, &mouse_ray.ray, true)
+            } {
+                if let Some(ref mut dist) = nearest_dist {
+                    if collision_distance < *dist {
+                        *dist = collision_distance;
+                        nearest_entity = Some(entity);
+                    }
+                } else {
+                    nearest_dist = Some(collision_distance);
+                    nearest_entity = Some(entity);
+                }
+            }
+        }
+        if nearest_entity != **hovered {
+            // something has changed
+
+            // if the hovered contains an entity, that entity is not hovered anymore
+            (**hovered).map(|e| {
```

Comment:
you're not mapping `hovered` onto anything, i'd suggest to use `if let` instead.